### ISSUE TITLE
PassengerAgent System Policy: deny mach-priv-task-port

### ISSUE THREAD

**Issue by jamgold** [phusion/passenger#1692]
Hello,

Running Phusion Passenger 5.0.21 on OSX 10.11.1 El Capitan and the System log is full of error messages about PassengerAgent

```12/5/15 10:01:06.000 kernel[0]: Sandbox: PassengerAgent(44038) System Policy: deny(1) mach-priv-task-port 44046```

expanding the error shows following. Anybody know how to fix this?

```
PassengerAgent(44038) System Policy: deny mach-priv-task-port

Process:         PassengerAgent [44038]
Path:            /usr/local/Cellar/passenger/5.0.21/libexec/buildout/support-binaries/PassengerAgent
Load Address:    0x10bdd7000
Identifier:      PassengerAgent
Version:         ??? (???)
Code Type:       x86_64 (Native)
Parent Process:  PassengerAgent [44036]

Date/Time:       2015-12-05 09:53:26.014 -0800
OS Version:      Mac OS X 10.11.1 (15B42)
Report Version:  8
```
[4367774.zip](https://github.com/phusion/passenger/files/52909/4367774.zip)

**Comment by jamgold** [comment]
Upgraded to 5.0.22 but the problem persists.
```
kernel[0]: Sandbox: PassengerAgent(72384) System Policy: deny(1) mach-priv-task-port 6987
```

**Comment by jamgold** [comment]
One more update on this: I have deployed two nodejs apps via passenger

* ghost
* meteor app

only the meteor app exhibits the problem with PassengerAgent.

**Comment by jamgold** [comment]
I am now on 5.0.30 and this is still happening. Nobody else seeing this?

**Comment by CamJN** [comment]
Could you provide a bit more info about how you are running passenger, and your meteor app? Or better yet can you provide a minimal meteor app that reproduces the issue?



notes for myself:
This seems to be related to SIP restrictions on using `task_for_pid`. 

htop experienced a similar issue: https://github.com/hishamhm/htop/issues/449.
monit too: https://bitbucket.org/tildeslash/monit/issues/266 & https://bitbucket.org/tildeslash/monit/commits/2fbf24563b88

SIP reference: https://developer.apple.com/library/watchos/documentation/Security/Conceptual/System_Integrity_Protection_Guide/System_Integrity_Protection_Guide.pdf

**Comment by jamgold** [comment]
Thank you for responding. I am pretty sure this is related to SIP. Will provide more information and a minimal app to reproduce.

**Comment by OnixGH** [comment]
@jamgold any update?

**Comment by jamgold** [comment]
Sorry @OnixGH, totally dropped the ball. I am still seeing the log entries but everything else works, so I stopped looking at the logs. Not sure how to exactly provide a minimal app to reproduce without shutting down my apps running under Passenger right now.

**Comment by jamgold** [comment]
Quick update. For every kernel[0]: Sandbox: PassengerAgent(473) System Policy: allow(0) mach-priv-task-port message a log file is being created in /var/log/asl/AUX....

On my Mac quite a lot of files accumulated. In /var/log/asl/AUX.2017.07.06 alone 10819, and over all the AUX.2017 directories 77747 files.

I have attached the content of one of those files

[42467564.txt](https://github.com/phusion/passenger/files/1132091/42467564.txt)

**Comment by CamJN** [comment]
So that confirms that the source of the logs is the call to `task_for_pid` in `measureRealMemory` in `ProcessMetricsCollector.h`. However [according to Apple](https://developer.apple.com/library/content/documentation/Security/Conceptual/System_Integrity_Protection_Guide/RuntimeProtections/RuntimeProtections.html#//apple_ref/doc/uid/TP40016462-CH3-SW1) (and consistent with my experience) you should be able to call `task_for_pid` on your own processes. So perhaps SIP is mistakenly protecting the `/usr/local/*` sub-directories on your mac? 

You can try and check this with the following command:

`ls -@Rl /usr/local | fgrep -i rootless`

If there's any output then SIP is confused.

**Comment by jamgold** [comment]
Ran that command and no output

**Comment by CamJN** [comment]
Weird, I wonder why SIP thinks it needs to protect the process then.

**Comment by CamJN** [comment]
Does the line including `/usr/local` start with a `*` in `/System/Library/Sandbox/rootless.conf`?

**Comment by jamgold** [comment]
yes, here is the entry

*               /usr/local

my webroot is on a different disk, can that matter?

**Comment by CamJN** [comment]
It shouldn't matter, no. We are measuring the memory use of Passenger when we make that call, so it's the passenger process that's being interfered with by SIP.

**Comment by jamgold** [comment]
the /usr directory has com.apple.rootless

 l -@d /usr 
drwxr-xr-x@ 13 root  wheel  442 Nov 30  2016 /usr
	com.apple.FinderInfo	 32 
	com.apple.rootless	  0

**Comment by CamJN** [comment]
Yes, `/usr` is expected to be protected only `/usr/local` is provided to the user.

**Comment by CamJN** [comment]
Oh I just realized something, do you have SIP disabled? you can check with `csrutil status`.

**Comment by CamJN** [comment]
You used to get logs like:
`kernel[0]: Sandbox: PassengerAgent(72384) System Policy: deny(1) mach-priv-task-port 6987`
which would imply that SIP was preventing the call but your latest log says:
`kernel[0]: Sandbox: PassengerAgent(473) System Policy: allow(0) mach-priv-task-port`
which is logged every time `task_for_pid` is called if SIP is disabled.

**Comment by jamgold** [comment]
Damn.
 csrutil status
System Integrity Protection status: disabled.

Let me enable SIP and report back

**Comment by jamgold** [comment]
Unfortunately back to deny every 5 seconds

`7/7/17 14:25:15.000 kernel[0]: Sandbox: PassengerAgent(385) System Policy: deny(1) mach-priv-task-port`

**Comment by jamgold** [comment]
One more thing, I run Phusion Passenger via nginx (don't know if that matters)

**Comment by CamJN** [comment]
How do you start nginx? Is it manual or launchd or something else?

**Comment by jamgold** [comment]
Launched by launchd. Using Homebrew version of nginx

**Comment by CamJN** [comment]
And is it started as a LaunchDaemon at boot or as a LaunchAgent on login?

**Comment by jamgold** [comment]
LaunchDaemon at boot via `/Library/LaunchDaemons/homebrew.mxcl.nginx.plist`

**Comment by CamJN** [comment]
Unfortunately I still cannot reproduce when running Nginx & Passenger that way. I'll keep thinking about this, but it's quite a riddle at this point.

**Comment by jamgold** [comment]
I am very grateful you haven't given up yet :)

**Comment by CamJN** [comment]
Ok so one more thing to check as to why SIP thinks it needs to protect `PassengerAgent`: `xxd /usr/local/opt/passenger/libexec/buildout/support-binaries/PassengerAgent | fgrep __RESTRICT`

I'm not confident that will reveal anything but it's something we haven't tried yet.

**Comment by jamgold** [comment]
It does not find __RESTRICT in the output of xxd

**Comment by CamJN** [comment]
I forgot to ask you to check the launchd plist for rootless attrs, so please do that:

`ls -Ol@ /Library/LaunchDaemons/homebrew.mxcl.nginx.plist`

When you say your webroot is on another disk, is it network mounted or connected via usb/firewire/thunderbolt/etc? It shouldn't matter but we've already covered all the sane explanations.

Does the disk that your webroot is on contain another osx install?

Can you also check the meteor app: 

`ls -@Rl /Volumes/webroot_Volume/path/to/app | fgrep -i rootless`

Are you running the Meteor app as a bundled meteor-application or in dev mode? How was meteor installed? And where? Same for node.

Thanks.

**Comment by jamgold** [comment]
Hi
```
± ls -Ol@ /Library/LaunchDaemons/homebrew.mxcl.nginx.plist
-rw-r--r--  1 root  wheel  - 571 Dec 15  2015 /Library/LaunchDaemons/homebrew.mxcl.nginx.plist
```
The webroot is on two different partitions (internal disks), none has rootless set in the attributes. Most applications run bundled /Volumes/Home/Web (no OS), but one runs in dev mode on /Users (OS X boot). Meteor was installed the usual way in my $HOME. All apps use the node binary for the corresponding meteor version the app is in. 

One last tidbit I haven't mentioned so far is that this is on a 2006 MacPro, so it has a community boot shim to support El Capitan. PassengerAgent is the only process complaining about mach-priv-task-port

**Comment by CamJN** [comment]
Hmm can you point me at the boot shim? That sounds unusual enough to possibly be related.

**Comment by jamgold** [comment]
I am using this alternative [macosxbootloader](https://github.com/Piker-Alpha/macosxbootloader) that allows me to run El Capitan. There is also [this](https://forums.macrumors.com/threads/2006-2007-mac-pro-1-1-2-1-and-os-x-el-capitan.1890435/) rather lengthy forum topic with instructions.

The system disk is a SSD and I am running a unmodified PC Titanium GFX card. Computer is humming along nicely. In the meantime I wrote a script that deletes ~2GB of ASL logs from /var/log every day

**Comment by CamJN** [comment]
Hmm you might be able to avoid the logs being written to disk using the ASL rules from the htop thread. Paraphrased below:

Add the following to `/etc/asl.conf`, above the line: `Rules for /var/log/system.log`
```
? [= Sender kernel] [= Facility kern] [N= Level 5] [Z= Message allow(0) mach-priv-task-port] ignore
? [= Sender sandboxd] [= Facility com.apple.sandbox] [N= Level 5] [Z= Message allow mach-priv-task-port] ignore
```
Activate with: `sudo pkill -HUP syslogd`

If you've disabled SIP since last time you enabled it, then the pattern will need to be adjusted to match the deny messages not the allow messages (or just block both).

But I'd really like to figure out what the root cause is here just in case we can do something to avoid it in Passenger.

**Comment by jamgold** [comment]
Cool, I will look into that. Let me know what (if any) more information I can provide. I just noticed I have several node crash reports in /Library/Logs/DiagnosticReports, but they don't line up with the frequency of the PassengerAgent log entries by a long shot.

**Comment by CamJN** [comment]
Feel free to just throw anything in here, who knows what might be relevant.

**Comment by CamJN** [comment]
So I don't see anything obviously wrong with the SIP handling in the bootloader, but I'm not an expert on that topic. https://github.com/Piker-Alpha/macosxbootloader/blob/1abba11e3b792dd29c2a08c410c122efd1d19e98/src/boot/BootArgs.cpp#L791-L896

Can you check if the executable that is running the meteor app has any special flags on it? You said earlier that the meteor app is bundled so that would mean it's being run by a node executable. You can just: `ls -Ol@` it. And please check for a `__RESTRICT` section too, just like before with something like `xxd /path/to/node | fgrep __RESTRICT`.

**Comment by jamgold** [comment]
No special flags on any of the node binaries on my system, and xxd doesn't find __RESTRICT either. Also checked my HomeBrew passenger and nginx directory for flags and __RESTRICT. Nothing.

Second interesting thing is that editing /etc/asl.conf does nothing to prevent logging of the sandboxd messages.

**Comment by jamgold** [comment]
Just read this [here](https://developer.apple.com/library/content/documentation/Security/Conceptual/System_Integrity_Protection_Guide/RuntimeProtections/RuntimeProtections.html#//apple_ref/doc/uid/TP40016462-CH3-SW1) and wonder if PassengerAgent is signed or if I am missing a particular System Certificate?
> When a process is started, the kernel checks to see whether the main executable is protected on disk or is signed with an special system entitlement. If either is true, then a flag is set to denote that it is protected against modification. Any attempt to attach to a protected process is denied by the kernel.

**Comment by CamJN** [comment]
PassengerAgent shouldn't be signed in that way, as only apple can provide the system entitlement described in that document. That's the basis for how SIP is supposed to work. Only Apple has the ability to create executables with protected entitlements, and only executables with those entitlements can modify SIP protected files & directories.

But why SIP is protecting PassengerAgent or the node executable is the question. Have you run disk utility's first aid on the disks involved in your setup? Perhaps there's been a corruption and an inode number got duplicated somehow.

**Comment by jamgold** [comment]
Thanks for that clarification. Will run disk utility in a sec. 

Going through my HomeBrew setup I just upgraded to Passenger 5.1.6 and notice that it comes with its own nginx-1.12.1 binary.

I am running nginx from HomeBrew as well, alas version 1.10.1 and it doesn't allow me to upgrade due to 
```
Error: nginx contains conflicting version recursive dependencies:
  openssl, openssl@1.1
View these with `brew deps --tree nginx`.
```

I am going to passenger-install-nginx and see if that changes anything

**Comment by CamJN** [comment]
The nginx binary that ships with the Passenger homebrew bottle/formula is there for Passenger Standalone to use. The nginx formula with the `--with-passenger` flag is how homebrew handles compiling passenger into nginx.

The conflicting dependencies for the nginx formula sounds like one or more parts of homebrew didn't update correctly. You might want to try the following:

`cd "$(brew --repo)" && git fetch && git reset --hard origin/master && cd Library/Taps/homebrew/homebrew-core && git fetch && git reset --hard origin/master`

That will forcibly update homebrew to the latest commit.

**Comment by CamJN** [comment]
Did homebrew build Passenger from scratch or download a bottle?

**Comment by jamgold** [comment]
Good question, I don't know and can't check right now, since my MacPro flamed out on Thursday and I need to wait for the replacement to be shipped.

**Comment by CamJN** [comment]
Ah, sorry to hear it.

**Comment by jamgold** [comment]
Resurrected my MacPro. To answer your question

`passenger: stable 5.1.7 (bottled), HEAD`

**Comment by CamJN** [comment]
Ok then, `brew install nginx --with-passenger` is definitely the approach I'd recommend for getting passenger integrated with nginx.

**Comment by jamgold** [comment]
OK, installed nginx 1.12.1 --with-passenger, same problem

**Comment by jamgold** [comment]
Quick question: after installing with `brew install nginx --with-passenger`, do I need to change the nginx config? Currently it still contains the line
```
passenger_root /usr/local/opt/passenger/libexec/src/ruby_supportlib/phusion_passenger/locations.ini;
```

**Comment by CamJN** [comment]
That line is correct for a homebrew install.

**Comment by jamgold** [comment]
I just found [this](http://attilathedud.me/mac-os-x-el-capitan-10-11-and-task_for_pid/) blog entry that explains how to enable task_for_pid in a binary (PassengerAgent), unfortunately I don't understand the half of it

**Comment by jamgold** [comment]
Through some testing I just found out that just one of my passenger node apps was causing these errors. It was to a Meteor app running in development mode. I changed my nginx config to simply reverse proxy to a manually started Meteor environment and all the syslog messages of PassengerAgent have stopped.

**Comment by CamJN** [comment]
I've got a Meteor app running in dev mode under passenger 5.1.7 on my el-cap box. So far no messages, but I'll leave it running.

**Comment by jamgold** [comment]
That is interesting. Obviously passenger via Homebrew and nginx?

**Comment by CamJN** [comment]
Yup, and meteor is installed with the normal curl-into-bash method.

**Comment by jamgold** [comment]
And all else you did was `brew install nginx --with-passenger` and you get no task_for_pid errors from PassengerAgent? And you didn't mess with any SIP/rootless settings?

**Comment by CamJN** [comment]
Yeah, just like that. It's a real head-scratcher.

**Comment by jamgold** [comment]
It's gotta be related to my MacPro hacked boot.efi, which seems the only thing different. I will try to reproduce with a small Meteor app.

**Comment by sfoop** [comment]
My Late 2013 Mac Pro is spewing log messages for this every 5 seconds and I've just started looking into it, so it's not just your boot.efi Mac Pro. I'm using the system apache and I've built passenger myself (v5.1.5, source tarball method), it's hosting a Ruby on Rails app.

    sandboxd[148]: ([84419]) PassengerAgent(84419) System Policy: deny mach-priv-task-port
    kernel[0]: Sandbox: PassengerAgent(84419) System Policy: deny(1) mach-priv-task-port

The pids are high because I just restarted apache.

**Comment by CamJN** [comment]
@sfoop can you provide the following info:

- os version
- which ruby your app is running with

**Comment by sfoop** [comment]
El Capitan 10.11.6 with system ruby 2.0.0p648 (/usr/bin/ruby).

**Comment by CamJN** [comment]
This is expected for the system ruby: 
```
ls -Ol@ /usr/bin/ruby
-r-xr-xr-x  1 root  wheel  restricted,compressed   51K 28 Apr 18:29 /usr/bin/ruby
```

**Comment by sfoop** [comment]
So the log messages are expected? Or do you want this info:

    ls -Ol@ /usr/bin/ruby
    8 -r-xr-xr-x  1 root  wheel  restricted,compressed 42864 May  4  2016 /usr/bin/ruby

**Comment by CamJN** [comment]
When you target a SIP protected executable (such as the system ruby) with `task_for_pid` it is expected that the function call will fail and a message will be written to the log. What makes @jamgold's case different is that his executable should not be SIP protected, as it came from the meteor folks not Apple.

**Comment by sfoop** [comment]
Is there a way to silence this?

**Comment by CamJN** [comment]
It was suggested in another project's issue that it may be possible to filter the lines out of the log; I summarized the info here: https://github.com/phusion/passenger/issues/1692#issuecomment-316457058 but it didn't work for @jamgold so I'm not confident that the info is good.

**Comment by sfoop** [comment]
Ok, thanks for your help!

**Comment by CamJN** [comment]
Is this still an issue?

**Comment by jamgold** [comment]
This is still an issue for me, now running under nginx. 

But it only happens when using  `passenger_app_env development`. This affects only one app I am developing, so I disabled passenger for that app and just use nginx reverse proxying instead
